### PR TITLE
Accelerator refactor sharded tpu fix

### PR DESCRIPTION
**Pull request by lezwon**
## What does this PR do?

Fixes # (issue) <- this [links related issue to this PR](https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [ ] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 1]
```diff
@@ -173,16 +173,16 @@ def __load_weights_on_main_process(self) -> None:
...
 
-        self.lightning_module = model
+        self._model = model
 
     @property
     def xmp_spawn_kwargs(self):
         return {
-            "args": (self.lightning_module, trainer, self.mp_queue),
-            "nproc": len(self.parallel_devices),
+            "args": (self.lightning_module.trainer, ),
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:184]: I recently re-added the queue here as input, as it seems necessary to pass it into the process, 
the self.new_process also has to be adapted

**Review by awaelchli — COMMENTED** [review]
Thanks a lot for helping with this!

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 1 above)
Comment (lezwon) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:184]: yep. `self.new_process` only accepts `trainer` arg as of now.

File: pytorch_lightning/accelerators/accelerator_connector.py [hunk 2]
```diff
@@ -364,7 +364,10 @@ def select_training_type_plugin(self):
         elif self.use_horovod:
             plugin = HorovodPlugin(parallel_devices=self.parallel_devices)
         elif self.on_tpu:
-            plugin = SingleTPUPlugin(self.tpu_id)
+            if isinstance(self.tpu_cores, list):
+                plugin = SingleTPUPlugin(self.tpu_id)
+            else:
```
[the remark this answers is not in the document]
Comment (lezwon) [reply pytorch_lightning/accelerators/accelerator_connector.py:369]: `[3,4,6]` is not supported. Only a single TPU core id is.

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 3]
```diff
@@ -231,13 +231,13 @@ def reduce(self, output, group: Optional[Any] = None, reduce_op: Optional[Union[
         return output
 
     def training_step(self, *args, **kwargs):
-        return self.model(*args, **kwargs)
+        return self.lightning_module.training_step(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (lezwon) [reply pytorch_lightning/plugins/training_type/ddp_spawn.py:234]: I was facing an error where in the forward method only accepted one arg : `batch`, but two were provided to it. i.e `batch` and `batch_idx`. Changed it just to get the test working.

**Comment by lezwon** [comment]
@justusschock @awaelchli Also im not sure if this line is ever called https://github.com/PyTorchLightning/pytorch-lightning/blob/994ac82549dc849f6a56a82e9cfa86b0a5a0a690/pytorch_lightning/accelerators/accelerator_connector.py#L342

File: pytorch_lightning/plugins/training_type/ddp_spawn.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (justusschock) [reply pytorch_lightning/plugins/training_type/ddp_spawn.py:234]: Yes, but please inside the TPU spawn class then :)

**Comment by justusschock** [comment]
@lezwon why should this line be never called? This should be called when we train on multiple tpu cores

**Comment by lezwon** [comment]
@justusschock it should be called. But I don't think it is. When training on tpu I don't think the Ddp variable is set.

**Comment by SeanNaren** [comment]
I think this property need to contain TPU as well, however not sure it make sense (`use_ddp` should be selected for TPUs if `tpu_cores > 1`)

https://github.com/PyTorchLightning/pytorch-lightning/blob/994ac82549dc849f6a56a82e9cfa86b0a5a0a690/pytorch_lightning/accelerators/accelerator_connector.py#L239-L243

**Pull request closed without merging**